File: TS/hellointerview/sliding-window/maxScore.ts
function maxScore(cards: number[], k: number): number {
  const total = cards.reduce((acc, x) => acc + x, 0);
  if (k >= cards.length) {
    return total;
  }

  let state = 0;
  let maxPoints = 0;
  let start = 0;

  for (let end = 0; end < cards.length; end++) {
    state += cards[end];

    if (end - start + 1 === cards.length - k) {
      maxPoints = Math.max(maxPoints, total - state);
      state -= cards[start];
      start++;
    }
  }

  return maxPoints;
}

File: Java/hellointerview/02-sliding-window/maxScore.java
import java.util.Arrays;

public class Solution {
  public int maxScore(int[] cards, int k) {
    int total = Arrays.stream(cards).sum();

    if (k >= cards.length) {
      return total;
    }

    int windowSum = 0;
    int result = 0;
    int start = 0;

    for (int end = 0; end < cards.length; end++) {
      windowSum += cards[end];
      if (end - start + 1 == cards.length - k) {
        result = Math.max(result, windowSum);
        windowSum -= cards[start];
        start++;
      }
    }

    return result;
  }
}
